lightningd/invoice.c

```diff
@@ -1441,3 +1475,162 @@ static const struct json_command decodepay_command = {
 	"Decode {bolt11}, using {description} if necessary"
 };
 AUTODATA(json_command, &decodepay_command);
+
+/* If we fail because it exists, we also return the clashing invoice */
+static struct command_result *fail_exists(struct command *cmd,
+					  const struct json_escape *label)
+{
+	struct json_stream *data;
+	struct invoice invoice;
+	struct wallet *wallet = cmd->ld->wallet;
+
+	data = json_stream_fail(cmd, INVOICE_LABEL_ALREADY_EXISTS,
+				"Duplicate label");
+	if (!wallet_invoice_find_by_label(wallet, &invoice, label))
+		fatal("Duplicate invoice %s not found any more?",
+		      label->s);
+
+	json_add_invoice(data, wallet_invoice_details(cmd, wallet, invoice));
+	json_object_end(data);
+
+	return command_failed(cmd, data);
+}
+
+static struct command_result *json_createinvoice(struct command *cmd,
+						 const char *buffer,
+						 const jsmntok_t *obj UNNEEDED,
+						 const jsmntok_t *params)
+{
+	const char *invstring;
+	struct json_escape *label;
+	struct preimage *preimage;
+	struct invoice invoice;
+	struct sha256 payment_hash;
+	struct json_stream *response;
+	struct bolt11 *b11;
+	struct sha256 hash;
+	u5 *sig;
+	bool have_n;
+	char *fail;
+
+	if (!param(cmd, buffer, params,
+		   p_req("invstring", param_string, &invstring),
+		   p_req("preimage", param_preimage, &preimage),
+		   p_req("label", param_label, &label),
+		   NULL))
+		return command_param_failed();
+
+	sha256(&payment_hash, preimage, sizeof(*preimage));
+	b11 = bolt11_decode_nosig(cmd, invstring, cmd->ld->our_features,
+				  NULL, chainparams, &hash, &sig, &have_n,
+				  &fail);
+	if (b11) {
+		/* This adds the signature */
+		char *b11enc = bolt11_encode(cmd, b11, have_n,
+					     hsm_sign_b11, cmd->ld);
+
+		if (!b11->description)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Missing description in invoice");
+
+		if (!b11->expiry)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Missing expiry in invoice");
+
+		if (!wallet_invoice_create(cmd->ld->wallet,
+					   &invoice,
+					   b11->msat,
+					   label,
+					   b11->expiry,
+					   b11enc,
+					   b11->description,
+					   b11->features,
+					   preimage,
```

Comment:
I guess it not being verified that the preimage matches the hash is a form of my desired situation above, but if that's intended we should make sure that we handle preimage / hash mismatches correctly.

Reply:
No, I imagine if we don't get an internal error, we'll at least lose the channel when we serve our peer an incorrect preimage.  I will enforce it here with a check.